ci: build pull requests in copr

pbrezina opened the pull request:
New copr project will be created for each pull request under the sssd
group. The project will be automatically deleted after 60 days or after
the pull request is closed.

The workflow also produces comment on the pull request, there may
be a little race condition between the time where the comment is
already visible and the time when the project is created, but it is
not long. The reason for this is that the comment is created before
the project is built in order for it to be always on the top.

The workflow works like this:
- submit comment to the pull request
- build tarball and spec file (rvn is sssd-pr#prid-#runid)
- build srpm
- upload srpm as an artifact to the job
- create copr project @sssd/pr#prid if not exist
- cancel previous pending/running builds if there are any
- build package in centos-* and fedora-*

The workflow fails if build in any chroot fails.
The workflow succeeds if all chroots are successful.

---

You may want to disable running all or selected actions in your forks to prevent workflows run in your repository.

pbrezina commented:
Note that the workflow will be picked up by GitHub after it is pushed to master. It's not yet going to run on this PR itself.

alexey-tikhonov commented:
> New project will be created for each pull request under the @SSSD name space

What is the purpose of those projects? How is it used / why is it needed to "build pull requests in copr"?

Does Github support nested namespaces to avoid pollution of top level namespace?

pbrezina commented:
> > New project will be created for each pull request under the @SSSD name space
> 
> What is the purpose of those projects? How is it used / why is it needed to "build pull requests in copr"?

See authselect for an example: https://github.com/authselect/authselect/pull/288 or https://copr.fedorainfracloud.org/groups/g/authselect/coprs/

It will build rpms in multiple chroots (centos and fedora) and makes it easily available via copr for test purpose.

> Does Github support nested namespaces to avoid pollution of top level namespace?

Unfortunately, no.

alexey-tikhonov commented:
> > > New project will be created for each pull request under the @SSSD name space
> > 
> > What is the purpose of those projects? How is it used / why is it needed to "build pull requests in copr"?
> 
> See authselect for an example: [authselect/authselect#288](https://github.com/authselect/authselect/pull/288) or https://copr.fedorainfracloud.org/groups/g/authselect/coprs/

I understand (and welcome) the plan to do Fedora builds automatically.
What I miss: why separate Github project is required - why this build can't be done via runner within PR?

> > Does Github support nested namespaces to avoid pollution of top level namespace?
> 
> Unfortunately, no. I can move the actions to my personal repo if needed.

I'm somewhat worried because we usually tend to have ~20 open PRs.
And some PRs *probably* don't need a build at all. For example: multihost test updates, translations...

alexey-tikhonov commented:
> See authselect for an example: [authselect/authselect#288](https://github.com/authselect/authselect/pull/288) or https://copr.fedorainfracloud.org/groups/g/authselect/coprs/

Btw, it seems there is no repo for this open PR nder https://github.com/authselect ?

pbrezina commented:
> > > > New project will be created for each pull request under the @SSSD name space
> > > 
> > > 
> > > What is the purpose of those projects? How is it used / why is it needed to "build pull requests in copr"?
> > 
> > 
> > See authselect for an example: [authselect/authselect#288](https://github.com/authselect/authselect/pull/288) or https://copr.fedorainfracloud.org/groups/g/authselect/coprs/
> 
> I understand (and welcome) the plan to do Fedora builds automatically. What I miss: why separate Github project is required - why this build can't be done via runner within PR?

It is not GitHub project, but copr project. I see that github automatically converted `@sssd` from the commit message to @SSSD and that might have caused some confusion.

- it will create a source rpm and make it an artifact of the github workflow run (it will be available to download when you click on details)
- it creates project in copr `@sssd/pr1234` with centos and fedora chroots enabled
- builds the pr
- the workflow will fail if any builds fails

pbrezina commented:
> > Does Github support nested namespaces to avoid pollution of top level namespace?
> 
> Unfortunately, no. I can move the actions to my personal repo if needed.

I meant that I created three actions https://github.com/SSSD/action-copr-delete https://github.com/SSSD/action-copr-build https://github.com/SSSD/action-build-srpm that can be used by other projects as well.

alexey-tikhonov commented:
> > > > > New project will be created for each pull request under the @SSSD name space
> > > > What is the purpose of those projects? How is it used / why is it needed to "build pull requests in copr"?
> > > See authselect for an example: [authselect/authselect#288](https://github.com/authselect/authselect/pull/288) or https://copr.fedorainfracloud.org/groups/g/authselect/coprs/
> > I understand (and welcome) the plan to do Fedora builds automatically. What I miss: why separate Github project is required - why this build can't be done via runner within PR?
> It is not GitHub project, but copr project. I see that github automatically converted `@sssd` from the commit message to @SSSD and that might have caused some confusion.

Yeap, exactly.
Thanks for clarification.

pbrezina commented:
I updated the commit message and PR description.

Note that centos currently fails due to missing packages, I'm looking into it.

justin-stephenson commented:
Is https://github.com/SSSD/action-build-srpm needed if we already have https://github.com/SSSD/sssd/blob/master/contrib/fedora/make_srpm.sh ?

>> You may want to disable running all or selected actions in your forks to prevent workflows run in your repository.

 How about adding `if: github.repository_owner == 'SSSD'` to have this workflow skipped on forks, or do you see some benefit in allowing it to run on forks?

pbrezina commented:
> Is https://github.com/SSSD/action-build-srpm needed if we already have https://github.com/SSSD/sssd/blob/master/contrib/fedora/make_srpm.sh ?

I guess it could be modified to take custom version and release string. But at the same time, I wanted to make it more reusable.

> > > You may want to disable running all or selected actions in your forks to prevent workflows run in your repository.
> 
> How about adding `if: github.repository_owner == 'SSSD'` to have this workflow skipped on forks, or do you see some benefit in allowing it to run on forks?

That way you wouldn't be able to test it, unless you modify the workflow. But perhaps it makes sense here.

justin-stephenson commented:
> > Is https://github.com/SSSD/action-build-srpm needed if we already have https://github.com/SSSD/sssd/blob/master/contrib/fedora/make_srpm.sh ?
> 
> I guess it could be modified to take custom version and release string. But at the same time, I wanted to make it more reusable.

I actually need this make_srpm.sh functionality for my internal covscan work, but we can leave the action if it could be useful going forward.
> 
> > > > You may want to disable running all or selected actions in your forks to prevent workflows run in your repository.
> > 
> > 
> > How about adding `if: github.repository_owner == 'SSSD'` to have this workflow skipped on forks, or do you see some benefit in allowing it to run on forks?
> 
> That way you wouldn't be able to test it, unless you modify the workflow. But perhaps it makes sense here.

Okay, it makes more sense to give people the option of disabling it.

Lastly, will deleting the COPR build after the PR is closed lead to problems for QE to take advantage of using the COPR build to do their part?

pbrezina commented:
> > > Is https://github.com/SSSD/action-build-srpm needed if we already have https://github.com/SSSD/sssd/blob/master/contrib/fedora/make_srpm.sh ?
> > 
> > 
> > I guess it could be modified to take custom version and release string. But at the same time, I wanted to make it more reusable.
> 
> I actually need this make_srpm.sh functionality for my internal covscan work, but we can leave the action if it could be useful going forward.
> 
> > > > > You may want to disable running all or selected actions in your forks to prevent workflows run in your repository.
> > > 
> > > 
> > > How about adding `if: github.repository_owner == 'SSSD'` to have this workflow skipped on forks, or do you see some benefit in allowing it to run on forks?
> > 
> > 
> > That way you wouldn't be able to test it, unless you modify the workflow. But perhaps it makes sense here.
> 
> Okay, it makes more sense to give people the option of disabling it.
> 
> Lastly, will deleting the COPR build after the PR is closed lead to problems for QE to take advantage of using the COPR build to do their part?

PR copr will be deleted, but the changes will be in nightly copr: https://copr.fedorainfracloud.org/coprs/g/sssd/nightly/

justin-stephenson commented:
Thanks, Ack from my side.

pbrezina commented:
Let's postpone this till the rawhide and centos builds are fixed.

pbrezina commented:
It needs some changes to make it work on centos builds. This copr change needs to be pushed first: https://pagure.io/copr/copr/pull-request/2072

pbrezina commented:
> > Is https://github.com/SSSD/action-build-srpm needed if we already have https://github.com/SSSD/sssd/blob/master/contrib/fedora/make_srpm.sh ?
> 
> I guess it could be modified to take custom version and release string. But at the same time, I wanted to make it more reusable.
> 
> > > > You may want to disable running all or selected actions in your forks to prevent workflows run in your repository.
> > 
> > 
> > How about adding `if: github.repository_owner == 'SSSD'` to have this workflow skipped on forks, or do you see some benefit in allowing it to run on forks?
> 
> That way you wouldn't be able to test it, unless you modify the workflow. But perhaps it makes sense here.

More importantly, you should not run any code from forked repository when in `pull_request_target` trigger, since it has access to secrets that could be stolen.

pbrezina commented:
The copr pull request was merged, but it will not be released any time soon, so there is no reason to wait.

I have reworked the PR quite a lot. Now there is now comment (it can be easily added again, but I didn't like it a lot) and there is one job for each chroot so its easy to track which chroot succeed or fail.

justin-stephenson commented on .github/workflows/copr_build.yml:
```diff
@@ -0,0 +1,92 @@
+        repository: ${{ github.event.pull_request.head.repo.full_name }}
+    - name: Build source rpm
+      id: srpm
+      uses: ./.github/actions/build-srpm
```
IMO it is a bit confusing to call `.github/actions/build-srpm`, which acts as a wrapper for `https://github.com/SSSD/action-build-srpm`

It would be great if these steps could all be self-contained in `github.com/SSSD/action-build-srpm`, the fact that it is in SSSD github org implies it is for use only with SSSD. If that is not feasible then I propose to at least include `wrapper` somewhere in the name of  `.github/actions/build-srpm/action.yml`

justin-stephenson replied:
Okay that works.

Is it correct that actions created inside SSSD repo `.github/actions/` files are re-usable for internal use by the SSSD project only, and `github.com/SSSD/action*` for external/generic use?

pbrezina replied:
Yes. Even .github/actions can be reused outside sssd, but they are not meant to.

justin-stephenson commented on .github/workflows/copr_build.yml:
```diff
@@ -0,0 +1,92 @@
+      chroots_json: ${{ steps.chroots.outputs.json }}
+    permissions:
+      contents: read
+    steps:
```
Could you add a newline between steps for readability? 

I would also like to have a comment added similar to the commit message which describes the high level steps. IMO it is not easy to discern this just from quickly reading this .yml file

pbrezina replied:
Sure, done.

justin-stephenson commented:
> The copr pull request was merged, but it will not be released any time soon, so there is no reason to wait.
> 
> I have reworked the PR quite a lot. Now there is now comment (it can be easily added again, but I didn't like it a lot) and there is one job for each chroot so its easy to track which chroot succeed or fail.

Can you link me to a successful run ?

justin-stephenson commented:
> > > I have reworked the PR quite a lot. Now there is now comment (it can be easily added again, but I didn't like it a lot)
> > 
> > 
> > `Now there is now comment?` Sorry I don't follow...
> 
> There is _no_ comment. Sorry for the typo :-)
> 
> > > The copr pull request was merged, but it will not be released any time soon, so there is no reason to wait.
> > > I have reworked the PR quite a lot. Now there is now comment (it can be easily added again, but I didn't like it a lot) and there is one job for each chroot so its easy to track which chroot succeed or fail.
> > 
> > 
> > Can you link me to a successful run ?
> 
> [pbrezina#27](https://github.com/pbrezina/sssd/pull/27) https://copr.fedorainfracloud.org/coprs/g/sssd/pr27/

Looks good to me, it would be nice if we could have some notification message like https://packit.dev/packit-service-pr.png but it is outside the scope of this PR

pbrezina commented:
> > > > I have reworked the PR quite a lot. Now there is now comment (it can be easily added again, but I didn't like it a lot)
> > > 
> > > 
> > > `Now there is now comment?` Sorry I don't follow...
> > 
> > 
> > There is _no_ comment. Sorry for the typo :-)
> > > > The copr pull request was merged, but it will not be released any time soon, so there is no reason to wait.
> > > > I have reworked the PR quite a lot. Now there is now comment (it can be easily added again, but I didn't like it a lot) and there is one job for each chroot so its easy to track which chroot succeed or fail.
> > > 
> > > 
> > > Can you link me to a successful run ?
> > 
> > 
> > [pbrezina#27](https://github.com/pbrezina/sssd/pull/27) https://copr.fedorainfracloud.org/coprs/g/sssd/pr27/
> 
> Looks good to me, it would be nice if we could have some notification message like https://packit.dev/packit-service-pr.png but it is outside the scope of this PR

Originally, there was a comment like this https://github.com/authselect/authselect/pull/295#issuecomment-1039268439 but the status banner does not update immediately due to github caching.

justin-stephenson commented:
Ack from me, thank you.

pbrezina commented:
There was an error in when deleting project, it works now. See the diff [here](https://github.com/SSSD/sssd/compare/c378f7df0d0e20b54019009727982c0201f436b0..11cc366eef5598709858b7222b5e6931cd026fd2).

pbrezina commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5972

* `master`
    * b2154298773e354952239abec942a981b6b9808a - ci: build pull requests in copr
